From 5c2c2a3120f5dcefa7d744dd1c4b2e8125ee5d21 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Sun, 12 Jul 2020 17:01:17 +0300
Subject: [PATCH] [SW-0] compiler: add __weak define

Add define for the "weak" compiler attribute. This is to shorten all the
__attribute__((weak)) directives to just __weak.
This helps with long long lines and aligns the code to what we have in
U-Boot.

Change-Id: Iffae246d5e8b18c3f942c34998c5480a312b91ee
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/common/efuse.c b/common/efuse.c
index 49951ca..3bece08 100644
--- a/common/efuse.c
+++ b/common/efuse.c
@@ -48,7 +48,7 @@
 	return 0;
 }
 
-__attribute__((weak)) int preboot_fuse_read(u32 word, u32 *val)
+__weak int preboot_fuse_read(u32 word, u32 *val)
 {
 	int offset, ret;
 	u32 read_val;
diff --git a/common/pll.c b/common/pll.c
index 0890c80..46a67f0 100644
--- a/common/pll.c
+++ b/common/pll.c
@@ -82,7 +82,7 @@
 	}
 }
 
-__attribute__((weak)) u32 hl_pll_get_freq(u32 pll_base, enum hl_pll_div_out div)
+__weak u32 hl_pll_get_freq(u32 pll_base, enum hl_pll_div_out div)
 {
 	u32 nr, nf, od, div_sel, factor, offset;
 	u64 out_freq = SYS_REF_CLOCK;
diff --git a/gaudi/common.c b/gaudi/common.c
index 58e1516..0113afa 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -69,7 +69,7 @@
 	regmap_writel(1, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
 }
 
-__attribute__((weak)) void preboot_init_pcie_phy(void)
+__weak void preboot_init_pcie_phy(void)
 {
 	/* Reset PCIe PHY */
 	regmap_writel(0, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
diff --git a/goya2/common.c b/goya2/common.c
index ce194ba..5fa3f6e 100644
--- a/goya2/common.c
+++ b/goya2/common.c
@@ -73,7 +73,7 @@
 	regmap_writel(1, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
 }
 
-__attribute__((weak)) void preboot_init_pcie_phy(void)
+__weak void preboot_init_pcie_phy(void)
 {
 	/* Reset PCIe PHY */
 	regmap_writel(0, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
@@ -223,7 +223,7 @@
  * Wait for Boot Manager to complete PCIe image execution
  * and return to idle state, so it won't initiate UART access anymore
  */
-__attribute__((weak)) void preboot_sync_btl(void)
+__weak void preboot_sync_btl(void)
 {
 	while (regmap_readl(mmPSOC_GLOBAL_CONF_BTM_FSM))
 		;
diff --git a/goya2_arc/common.c b/goya2_arc/common.c
index 1dda76b..8abebeb 100644
--- a/goya2_arc/common.c
+++ b/goya2_arc/common.c
@@ -49,7 +49,7 @@
 	regmap_writel(1, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
 }
 
-__attribute__((weak)) void preboot_init_pcie_phy(void)
+__weak void preboot_init_pcie_phy(void)
 {
 	/* Reset PCIe PHY */
 	regmap_writel(0, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
diff --git a/goya2_arc_gnu/common.c b/goya2_arc_gnu/common.c
index b119690..a309275 100644
--- a/goya2_arc_gnu/common.c
+++ b/goya2_arc_gnu/common.c
@@ -77,7 +77,7 @@
 	regmap_writel(1, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
 }
 
-__attribute__((weak)) void preboot_init_pcie_phy(void)
+__weak void preboot_init_pcie_phy(void)
 {
 	/* Reset PCIe PHY */
 	regmap_writel(0, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
diff --git a/include/preboot.h b/include/preboot.h
index 857ed13..de0212e 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -48,6 +48,9 @@
 #ifndef __packed
 #define __packed	__attribute__((packed))
 #endif
+#ifndef __weak
+#define __weak		__attribute__((weak))
+#endif
 
 typedef unsigned char u8;
 typedef unsigned int u32;
diff --git a/ppboot/ppboot.c b/ppboot/ppboot.c
index 15a1c9b..8aa738c 100644
--- a/ppboot/ppboot.c
+++ b/ppboot/ppboot.c
@@ -40,7 +40,7 @@
 	return image_pending;
 }
 
-__attribute__((weak)) u64 copy_preboot(void)
+__weak u64 copy_preboot(void)
 {
 	char *name = "PREBOOT";
 	struct boot_desc desc;
@@ -109,8 +109,8 @@
 	return false;
 }
 
-__attribute__((weak)) u32 preboot_clr_flops_reset_cause(void) { return 0; }
-__attribute__((weak)) void ppboot_log_waiting_for_pcie(void) {}
+__weak u32 preboot_clr_flops_reset_cause(void) { return 0; }
+__weak void ppboot_log_waiting_for_pcie(void) {}
 
 static void ppboot_pcie_init(void)
 {
@@ -141,7 +141,7 @@
 	regmap_writel(reg, mmCPU_CA53_CFG_ARM_RST_CONTROL);
 }
 
-__attribute__((weak)) void hl_init_pcie_phy_fw(void) {}
+__weak void hl_init_pcie_phy_fw(void) {}
 
 static inline void ppboot_version(void)
 {
diff --git a/ppboot/spi.c b/ppboot/spi.c
index c818647..74459ce 100644
--- a/ppboot/spi.c
+++ b/ppboot/spi.c
@@ -42,7 +42,7 @@
  * mmPSOC_GPIO serve 16 GPIOs
  * Note: CS0 is active low
  */
-__attribute__((weak)) void spi_ssi_cs0_get(unsigned int *mask,
+__weak void spi_ssi_cs0_get(unsigned int *mask,
 					   unsigned int *data_offset,
 					   unsigned int *out_en_offset)
 {
@@ -200,7 +200,7 @@
 	spi_release_bus();
 }
 
-__attribute__((weak)) void spi_base_get(unsigned int *regs)
+__weak void spi_base_get(unsigned int *regs)
 {
 	*regs = mmPSOC_SPI_CTRLR0;
 }
diff --git a/preboot/fit.c b/preboot/fit.c
index c821ac8..edd3804 100644
--- a/preboot/fit.c
+++ b/preboot/fit.c
@@ -30,7 +30,7 @@
 
 #define FIT_ALGO_PROP_EXPECTED_VALUE	"sha256,rsa3072"
 
-__attribute__((weak)) int fit_need_image_verification(void)
+__weak int fit_need_image_verification(void)
 {
 	return 0;
 }
diff --git a/preboot/module_pci.c b/preboot/module_pci.c
index dda1a08..0089231 100644
--- a/preboot/module_pci.c
+++ b/preboot/module_pci.c
@@ -22,7 +22,7 @@
 	.hot_rst_reg_val = ~1,
 };
 
-__attribute__((weak)) void pci_dev_specific_setup(void) {}
+__weak void pci_dev_specific_setup(void) {}
 
 void module_pci_setup(void)
 {
@@ -30,7 +30,7 @@
 		pci_dev_specific_setup();
 }
 
-__attribute__((weak)) enum cmd_check_res dev_specific_module_pci_check(void)
+__weak enum cmd_check_res dev_specific_module_pci_check(void)
 {
 	pci_sts.flr_reg_val = regmap_readl(mmPCIE_AUX_FLR_INT);
 	pci_sts.hot_rst_reg_val = regmap_readl(mmPCIE_CORE_HOT_RST) & BIT(0);
diff --git a/preboot/module_uart.c b/preboot/module_uart.c
index 16f3259..339b448 100644
--- a/preboot/module_uart.c
+++ b/preboot/module_uart.c
@@ -19,7 +19,7 @@
 	return FOUND_FAST_RETURN;
 }
 
-__attribute__((weak)) void hl_parse_uart(int val) {}
+__weak void hl_parse_uart(int val) {}
 
 void module_uart_run(void)
 {
diff --git a/preboot/thermal.c b/preboot/thermal.c
index 3cfadf3..e3143cc 100644
--- a/preboot/thermal.c
+++ b/preboot/thermal.c
@@ -36,7 +36,7 @@
 
 struct hl_ts_cfg hl_ts_data[HL_ASIC_TS_MAX_COUNT];
 
-__attribute__((weak)) void hl_ts_clk_set(u32 addr, unsigned int clk)
+__weak void hl_ts_clk_set(u32 addr, unsigned int clk)
 {
 	unsigned int synth_clk;
 	unsigned int val;
@@ -50,7 +50,7 @@
 
 }
 
-__attribute__((weak)) void hl_ts_sdif_set(u32 addr, bool use_calib)
+__weak void hl_ts_sdif_set(u32 addr, bool use_calib)
 {
 	udelay(100);
 	/* program sensor power up delay */
@@ -64,7 +64,7 @@
 }
 
 /* convert 12-bit HL_TMP register value to milliCelsius */
-__attribute__((weak)) long hl_tmp_temp_reg_to_mC(int val, int ts)
+__weak long hl_tmp_temp_reg_to_mC(int val, int ts)
 {
 	long tmp = (val & 0xfff);
 	u32 coef_a_g, coef_b_h, coef_j, coef_div, fclk;
@@ -89,7 +89,7 @@
 }
 
 /* convert milliCelsius to 12-bit HL_TMP register value */
-__attribute__((weak)) u32 hl_tmp_mC_to_temp_reg(long val, int ts)
+__weak u32 hl_tmp_mC_to_temp_reg(long val, int ts)
 {
 	u32 reg_val;
 	u32 coef_a_g, coef_b_h, coef_j, coef_div, fclk;
@@ -148,12 +148,12 @@
 	hl_ts_data[ts].hl_tmp_data.coef_b_h += efuse_b_coeff / 10;
 }
 
-__attribute__((weak)) int get_alarma_offset(void)
+__weak int get_alarma_offset(void)
 {
 	return HL_TS0_ALARMA_CFG;
 }
 
-__attribute__((weak)) int get_alarmb_offset(void)
+__weak int get_alarmb_offset(void)
 {
 	return HL_TS0_ALARMB_CFG;
 }
@@ -172,7 +172,7 @@
 	regmap_writel(reg, hl_ts_data[ts_num].base + get_alarmb_offset());
 }
 
-__attribute__((weak)) void hl_set_calib_coeff(int ts, bool calib_enabled,
+__weak void hl_set_calib_coeff(int ts, bool calib_enabled,
 		struct hl_tmp_opmode_info *calib_data)
 {
 	if (calib_enabled) {
@@ -193,27 +193,27 @@
 
 }
 
-__attribute__((weak)) int hl_ts_get_count(void)
+__weak int hl_ts_get_count(void)
 {
 	return HL_ASIC_TS_MAX_COUNT;
 }
 
-__attribute__((weak)) int get_sdif_offset(void)
+__weak int get_sdif_offset(void)
 {
 	return HL_TS_SDIF;
 }
 
-__attribute__((weak)) int hl_ts_pvt_init(int ts)
+__weak int hl_ts_pvt_init(int ts)
 {
 	return 0;
 }
 
-__attribute__((weak)) void hl_ts_data_init(void)
+__weak void hl_ts_data_init(void)
 {
 	memset(hl_ts_data, 0, sizeof(struct hl_ts_cfg) * HL_ASIC_TS_MAX_COUNT);
 }
 
-__attribute__((weak)) void hl_dev_ts_init(void)
+__weak void hl_dev_ts_init(void)
 {
 	u32 efuse_word, ts;
 
@@ -250,7 +250,7 @@
 	hl_dev_ts_init();
 }
 
-__attribute__((weak)) int get_sdif_data_offset(void)
+__weak int get_sdif_data_offset(void)
 {
 	return HL_TS_SDIF_DATA_OFFSET;
 }
